### PR TITLE
feat(cmd/influx/write): enhance the ability to import CSV files using csv2lp library

### LINKED ISSUE #17004: Add the ability to import a simple csv file via the influx cli
#17398 introduces CSV data to `influx write` command. As a next step, we should not require (and reduce the need) to change CSV data on input.

_Enhancements in CSV data processing_ 

- `--header` option in `influx write` command let your add annotation or header rows without changing the data on input (supplied via `--file` or stdin)
   - you can supply more `--header` options, new CSV rows will be prepended in the order as they appear on the command line, you can also supply more`--file` options, files follow headers
- the header row (i.e. the row that define column names) can also define column data types when supplied as `name|datatype`; for example `cpu|tag` defines a tag column named _cpu_
  - moreover, it can also specify a default value when supplied as `name|datatype|default`; for example, `count|long|0` defines a field column named _count_ of _long_ data type that will not skip the field if a column value empty, but uses '0' as the column value
- a new `#constant` annotation adds a constant column to the data, so you can set measurement, time, field or tag of every row you import independently on the data you passed in
   - the format of a constant annotation row is `#constant,datatype,name,value`, so you specify a supported datatype, a column name, and a constant value
   - `column` can be omitted for _dateTime_ or _measurement_ columns, so the annotation can be simply `#constant,measurement,cpu`
   - note that you can add constant annotations to existing data using `--header` option of `influx write` cli
- a time column can be of `dateTime:format` datatype to use a custom _format_ to parse column values
   - the format follows the layout described in https://golang.org/pkg/time layout, for example `dateTime:2006-01-02` parses 4-digit-year , '-' , 2-digit month ,'-' , 2 digit day of the month
   - the following formats are predefined:
      - `dateTime:RFC3339` format is 2006-01-02T15:04:05Z07:00
      - `dateTime:RFC3339Nano` format is 2006-01-02T15:04:05.999999999Z07:00
      - `dateTime:number` represent UTCs time since epoch in nanoseconds
   - if the time format includes a time zone, the parsed date time respects the time zone; otherwise the timezone dependends on the presence of `#timezone` annotation; if there is no #timezone annotation, UTC is used
      - `#timezone` annotation specifies time zone using an offset , which is either `+hhmm` or `-hhmm` or `Local` to use the local/computer time zone. Examples:  _#timezone,+0100_  _#timezone -0500_ _#timezone Local_
          
- a _double_ field column can be supplied with `format` as `double:format`
   - the `format`'s first character is used to separate integer and fractional part (usually `.` or `,`), second and next format's characters (such as as `, _`) are removed  from the column value, these removed characters are usually used to visually separate large numbers into groups
   - for example a Spanish locale value `3.494.826.157,123` is of `double:,.` type; the same `double` value is  _3494826157.123_
   - _long_ or _unsignedLong_ data types also support format, everything after and including a fraction character is ignored
   - note that you have to quote column delimiters whenever they appear in a CSV column value, for example:
      - `#constant,"double:,.",myColumn,"1.234,011"`
- a _boolean_ field column can be supplied with `format` as `boolean:truthy:falsy`
   - `truthy` are `falsy` are comma-separated lists of values that are separated by `:` , they can be empty to assume other values as truthy/falsy
   - a  `boolean` without format parses columns values that start with any of _tTyY1_ as `true` values, _fFnN0_ as `false` values and fails on other values
   - a column with an empty value is excluded in the protocol line, setup `#default` annotation for such a column define default value 
   - for example `boolean:sí,yes,ja,oui,ano,да:no,nein,non,ne,нет` 
- a CSV file can start with a line `sep=;` to inform about a character that is used to separate columns in data rows, by default `,` is used as a column separator
- `--skipHeader` flag skips the first line from the input data (the last file or stdin); the optional value of this flag specifies the number of lines to remove; for example `skipHeader=2` removes the first 2 lines
- `--debug` and `--continueOnError` options help with troubleshooting of CSV conversions
   - `--debug` prints (to stderr) metadata about columns that are used to create protocol lines out of CSV data rows
   - `--skipRowOnError` prints CSV data conversion error to stderr, skips the failing row and continues with CSV processing
      - any error stops the processing without this option specified

### PR DESCRIPTION
**Pull request by sranka**
Closes #17004 #17008

This PR enhances the way of how CSV data are processed by `influx write` so that even external CSV data can be written to InfluxDB. 
- It uses a newly introduced `csv2lp package` to convert CSV to the line protocol. The conversion relies solely upon the content on CSV data on input. It is well documented in https://github.com/influxdata/influxdb/tree/master/pkg/csv2lp/README.md, examples are therein as well.
- there are a few new flags in `influx write` that helps the user to import the existing CSV file without changing it 
   - the CSV file encoding can be specified
   - existing header rows can be skipped and new header/annotation rows can be provided instead in order to describe the conversion to line protocol
  - multiple CSV files can be provided so that additional header/annotation rows can be placed in a file   
  - the user can better debug/tune the CSV to line protocol conversion
```
      --debug                Log CSV columns to stderr before reading data rows
      --encoding string      Character encoding of input files or stdin (default "UTF-8")
  -f, --file stringArray     The path to the file to import
      --header stringArray   One or more header lines to prepend to input data
      --skipHeader int[=1]   Skip the first <n> rows from input data
      --skipRowOnError       Log CSV data errors to stderr and continue with CSV processing
```

- [ ] [CHANGELOG.md](https://github.com/influxdata/influxdb/blob/master/CHANGELOG.md) updated with a link to the PR (not the Issue)
- [x] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0-beta.3/)
- [x] Rebased/mergeable
- [x] Tests pass
- [x] Documentation updated or issue created (provide link to issue/pr)
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

**Comment by sranka** [comment]
This PR is a continuation of #17599 that had to be split into two parts. The first part is a csv2lp library that is already in master via #17764. This PR is the second part.

**Comment by sranka** [comment]
Related documentation issue: influxdata/docs-v2#1050

**Comment by russorat** [comment]
I've tested this locally and it works great! looking forward to seeing this merged!

File: cmd/influx/internal/is_character_device.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package internal
+
+import (
+	"io"
+	"os"
+)
+
+// IsCharacterDevice returns true if the supplied reader is a character device (a terminal)
+func IsCharacterDevice(reader io.Reader) bool {
```
Comment (jsteenb2) [inline cmd/influx/internal/is_character_device.go:9]: can you move this to the cmd/influx main pkg itself. Not seeing the reuse here

File: write/csvTable.go [hunk 2]
```diff
@@ -1,420 +0,0 @@
-package write
```
Comment (jsteenb2) [inline write/csvTable.go:1]: Can you help me understand why these `write/*` deletions are here?

File: cmd/influx/internal/is_character_device.go (hunk 1 above)
Comment (sranka) [reply cmd/influx/internal/is_character_device.go]: moved to write

File: write/csvTable.go (hunk 2 above)
Comment (sranka) [reply write/csvTable.go:1]: The conversion of CSV to line protocol was refactored and improved in https://github.com/influxdata/influxdb/tree/master/pkg/csv2lp  as requested by https://github.com/influxdata/influxdb/pull/17599#issuecomment-614174079

**Pull request merged**